GeosCore/kpp_standalone_interface.F90

```diff
@@ -0,0 +1,698 @@
+      INTEGER,        INTENT(IN)    :: KPP_TotSteps ! Total KPP integrator steps
+
+      TYPE(GrdState), INTENT(IN)    :: State_Grid                 ! Grid State object
+      TYPE(ChmState), INTENT(IN)    :: State_Chm                  ! Chemistry State object
+      TYPE(MetState), INTENT(IN)    :: State_Met                  ! Meteorology State object
+      TYPE(OptInput), INTENT(IN)    :: Input_Opt                  ! Input Options object
+      REAL(dp), INTENT(IN)          :: initC(NSPEC)               ! Initial concentrations
+      REAL(dp), INTENT(IN)          :: localRCONST(NREACT)        ! Rate constants
+      REAL(dp)                      :: initHvalue                 ! Initial timestep
+
+! !OPTIONAL INPUT PARAMETER
+      LOGICAL, OPTIONAL             :: FORCE_WRITE  ! Write even if not in an active cell
+      CHARACTER(LEN=*), OPTIONAL    :: CELL_NAME    ! Customize the name of this file 
+!
+! !AUXILLIARY LOCAL PARAMETERS (pass the aux bc Fortran doesn't have defaults for kwargs)
+      LOGICAL            :: FORCE_WRITE_AUX   ! Write even if not in an active cell
+      CHARACTER(LEN=255) :: CELL_NAME_AUX     ! Customize the name of this file 
+!
+! !OUTPUT PARAMETERS:
+!
+      INTEGER,        INTENT(OUT)   :: RC          ! Success or failure
+!
+! !LOCAL VARIABLES:
+      ! Integers
+      INTEGER                :: N          ! Loop index
+      INTEGER                :: IU_FILE    ! Available unit for writing
+      INTEGER                :: SpcID      ! Mapping from State_Chm and KPP
+      REAL(fp)               :: DT         ! Chemistry operator timestep
+      ! Strings
+      CHARACTER(LEN=255)     :: YYYYMMDD_hhmmz
+      CHARACTER(LEN=255)     :: level_string
+      CHARACTER(LEN=512)     :: errMsg
+      
+      ! Arrays
+      REAL(dp)               :: Vloc(NVAR), Aout(NREACT)  ! For KPP reaction rate diagnostics
+
+
+      ! Did a user want to write the chemical state even if not in an active cell?
+      IF ( PRESENT(FORCE_WRITE) ) THEN
+         FORCE_WRITE_AUX = FORCE_WRITE
+      ELSE
+         FORCE_WRITE_AUX = .FALSE.
+      END IF
+
+      ! Quit early if there's no writing to be done
+      IF (KPP_Standalone_YAML%Active_Cell .eq. .FALSE. .AND. FORCE_WRITE_AUX .eq. .FALSE.) THEN 
+         RETURN
+      END IF
+
+      ! Did the call include an optional cell name?
+      IF ( PRESENT(CELL_NAME) ) THEN
+         CELL_NAME_AUX = CELL_NAME
+      ELSE
+         CELL_NAME_AUX = ''
+      END IF
+
+      CALL Fun( V       = initC(1:NVAR),                                     &
+                F       = initC(NVAR+1:NSPEC),                               &
+                RCT     = localRCONST,                                       &
+                Vdot    = Vloc,                                              &
+                Aout    = Aout                                          )
+
+      DT = GET_TS_CHEM()
+
+      !========================================================================
```

Comment:
This section needs to be looked at. I noticed that if I run again and do not delete the .txt files then I get multiple entries:
```
48
48
48
===========================================================================

                  KPP Standalone Atmospheric Chemical State
File Description:
This file contains model output of the atmospheric chemical state
as simulated by the GEOS-Chem chemistry module in a 3D setting.
Each grid cell represents the chemical state of an individual location,
suitable for input into a separate KPP Standalone program which will
replicate the chemical evolution of that grid cell for mechanism analysis.
Note that the KPP Standalone will only use concentrations, rate constants,
and KPP-specific fields. All other fields are for reference. The first line
contains the number of lines in this header. If wanting to use this output
for other analysis, a Python class to read these fields is available by
request, contact Obin Sturm (psturm@usc.edu).
48
===========================================================================

===========================================================================
===========================================================================

                  KPP Standalone Atmospheric Chemical State
File Description:
This file contains model output of the atmospheric chemical state
                  KPP Standalone Atmospheric Chemical State
```
I also noticed there are several fort.# files (e.g. fort.11, fort.12, etc) produced in the run directory. This indicates a problem connecting units to files. The contents are always the tail end of Aout values e.g. this in fort.11:
```
A877,   0.3112714359949602E-36
A878,   0.9944338835160137E-37
A879,   0.3973099614377623E-29
A880,   0.5543315029509759E-28
A881,   0.6884754883032322E-27
A882,   0.1581490519236982E-27
A883,   0.5030724551051664E-27
A884,   0.5472045352680872E-25
A885,   0.2563475681034608E-28
A886,   0.1940313854622394E-25
A887,   0.6198814341972948E-26
A888,   0.1149428960227215E-25
A889,   0.1086602171784139E-25
A890,   0.1635476817635733E-25
A891,   0.9736320085422249E-28
A892,   0.2378067741933753E-24
A893,   0.2861694322248196E-13
A894,   0.1339118880205419E-14
```

Reply:
This is interesting and I haven't seen it on my end.  Are you testing this with the` #ifdef MODEL_GEOS` lines turned off around line 1280 in `fullchem_mod.F90` / are you testing this with GEOS-Chem 14.4.0 in an OMP environment?  

@christophkeller and I are wondering if this could be coming from the `IU_FILE = findFreeLUN() ` line in `kpp_standalone_interface.F90`. Could it be that this needs to be handled differently when inside of the fullchem OMP loop?

Reply:
>  Could it be that this needs to be handled differently when inside of the fullchem OMP loop?
My suspicion is yes. I am testing this in GC-Classic, with the MODEL_GEOS ifdef removed. I have not yet tried in GCHP but will next week.

Reply:
Hi Lizzie, we were wondering about the best way to fix this output bug.  I don't have much knowledge of OpenMP, it is arising from OpenMP, would adding an`!$omp critical` statement somewhere in this section be a potential approach?  

Currently, we don't have a way of testing this module with GC-Classic on our end, but would definitely like this module to be compatible with GC-Classic.

Reply:
Yes, I will take a look at this. Would you like me to do my testing on discover so that you have access?

Reply:
That would be great! Do you also have GC-Classic running on Discover?

Reply:
Yes, I am now testing this on discover. Interestingly I do not see the same issue there. No fort.* files are generated. I wonder if it's dependent on compiler.

My run directory and code are at ~ewlundgr/nb/GCC. I added a few commits on top of yours at ~ewlundgr/nb/GCC/code.gcc_14.4/src/GEOS-Chem which you are welcome to cherry-pick into your PR branch before merging. I wouldn't let the fort.* issue hold up the merge. This will not come into the standard GEOS-Chem model until the next version since we are freezing 14.4 next week.

Reply:
Great, thank you Lizzie! I looked through your commits, we seem to have done largely identical commits in parallel, because I added some commits earlier this week in response to your review.  I also added an OutputDir backup check in case a user-specified directory doesn't exist in this commit, and only write to the working directory if there is no OutputDir (which won't be the case except in GEOS): `https://github.com/GEOS-ESM/geos-chem/pull/9/commits/2075fe7c8f9536c1109479fb9f14efff4cec5a5e`

Reply:
Resolving for now to not hold up the merge